allennlp/service/predictors/semantic_role_labeler.py

```diff
@@ -41,53 +41,147 @@ def make_srl_string(words: List[str], tags: List[str]) -> str:
+        together, rather than separately).
+
+        Parameters
+        ----------
+        json : ``JsonDict``, required.
+            JSON that looks like ``{"sentence": "..."}``.
+
+        Returns
+        -------
+        instances : ``List[Instance]``
+            One instance per verb.
+        result_dict : ``JsonDict``
+            A dictionary containing the words of the sentence and the verbs extracted
+            by the Spacy POS tagger. These will be replaced in ``predict_json`` with the
+            SRL frame for the verb.
         """
-        sentence = inputs["sentence"]
-
+        sentence = json["sentence"]
         tokens = self._tokenizer.split_words(sentence)
         words = [token.text for token in tokens]
-        results: JsonDict = {"words": words, "verbs": []}
+        result_dict: JsonDict = {"words": words, "verbs": []}
+        instances: List[Instance] = []
         for i, word in enumerate(tokens):
             if word.pos_ == "VERB":
                 verb = word.text
+                result_dict["verbs"].append(verb)
                 verb_labels = [0 for _ in words]
                 verb_labels[i] = 1
                 instance = self._dataset_reader.text_to_instance(tokens, verb_labels)
-                output = self._model.forward_on_instance(instance, cuda_device)
-                tags = output['tags']
+                instances.append(instance)
+        return instances, result_dict
 
-                description = SemanticRoleLabelerPredictor.make_srl_string(words, tags)
+    @overrides
+    def predict_batch_json(self, inputs: List[JsonDict], cuda_device: int = -1) -> List[JsonDict]:
+        """
+        Expects JSON that looks like ``[{"sentence": "..."}, {"sentence": "..."}, ...]``
+        and returns JSON that looks like
 
+        .. code-block:: js
+
+            [
+                {"words": [...],
+                 "verbs": [
+                    {"verb": "...", "description": "...", "tags": [...]},
+                    ...
+                    {"verb": "...", "description": "...", "tags": [...]},
+                ]},
+                {"words": [...],
+                 "verbs": [
+                    {"verb": "...", "description": "...", "tags": [...]},
+                    ...
+                    {"verb": "...", "description": "...", "tags": [...]},
+                ]}
+            ]
+        """
+        # For SRL, we have more instances than sentences, but the user specified
+        # a batch size with respect to the number of sentences passed, so we respect
+        # that here by taking the batch size which we use to be the number of sentences
+        # we are given.
+        batch_size = len(inputs)
+        instances_per_sentence, return_dicts = zip(*[self._sentence_to_srl_instances(json)
+                                                     for json in inputs])
+        flattened_instances = sum(instances_per_sentence, [])
```

Comment:
I dislike this construct (and I believe it's quadratic, which probably doesn't matter much here but makes it a good idea to avoid it on principle)

```
flattened_instances = [instance for sentence_instances in instances_per_sentence for instance in sentence_instances]
```

Reply:
I went down a massive rabbit hole reading about list monads when I looked this up, but thanks, I wasn't aware this was so inefficient. I also didn't know that `list(itertools.chain.from_iterable(some_list))` is actually _faster_ than the nested list comprehensions. weird.